README.md
[![Download](https://api.bintray.com/packages/sczerwinski/android/tiles-layout/images/download.svg)](https://bintray.com/sczerwinski/android/tiles-layout/_latestVersion)
[![Anko Download](https://img.shields.io/badge/Download-Anko-blue.svg)](https://bintray.com/sczerwinski/android/tiles-layout-anko/_latestVersion)

# Static Tiles Layout For Android And Anko

## Gradle Build Configuration

```gradle
dependencies {
    compile 'it.czerwinski.android:tiles-layout:0.1'
    compile 'it.czerwinski.android:tiles-layout-anko:0.1' // with Kotlin and Anko
}
```

## Usage

### XML layout

```xml
<it.czerwinski.android.view.layout.tiles.TilesLayout
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    android:paddingLeft="16dp"
    android:paddingRight="16dp"
    android:paddingTop="8dp"
    android:paddingBottom="8dp"
    app:tilesLayout_colsCount="2"
    app:tilesLayout_rowsCount="3"
    app:tilesLayout_innerHorizontalPadding="24dp"
    app:tilesLayout_innerVerticalPadding="12dp">

    <Button
        android:text="Button 1"
        android:layout_width="match_parent"
        android:layout_height="match_parent"/>

    <Button
        android:text="Button 2"
        android:layout_width="match_parent"
        android:layout_height="match_parent"/>

    <Button
        android:text="Button 3"
        android:layout_width="match_parent"
        android:layout_height="match_parent"/>

    <Button
        android:text="Button 4"
        android:layout_width="match_parent"
        android:layout_height="match_parent"/>

    <Button
        android:text="Button 5"
        android:layout_width="match_parent"
        android:layout_height="match_parent"/>

    <Button
        android:text="Button 6"
        android:layout_width="match_parent"
        android:layout_height="match_parent"/>

</it.czerwinski.android.view.layout.tiles.TilesLayout>
```

### Anko

```kotlin
tilesLayout(cols = 2, rows = 3) {
    horizontalPadding = dip(16)
    verticalPadding = dip(8)
    innerHorizontalPadding = dip(24)
    innerVerticalPadding = dip(12)

    button("Button 1")
    button("Button 2")
    button("Button 3")
    button("Button 4")
    button("Button 5")
    button("Button 6")
}
```


RELEASE-NOTES.md
# Static Tiles Layout For Android And Anko – Release Notes

## Version 0.1

* **Git tag:** tiles-layout-0.1

### New features

* Implementation of static tiles layout.


tiles-layout-anko/src/androidTest/kotlin/it/czerwinski/android/view/layout/tiles/anko/AnkoTilesLayoutTest.kt
package it.czerwinski.android.view.layout.tiles.anko

import android.app.Activity
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.matcher.ViewMatchers.*
import android.support.test.filters.LargeTest
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.view.View
import it.czerwinski.android.view.layout.tiles.TilesLayout
import org.hamcrest.Matchers.*
import org.jetbrains.anko.dip
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@LargeTest
@RunWith(AndroidJUnit4::class)
class AnkoTilesLayoutTest {

	@Rule
	@JvmField
	val testRule = ActivityTestRule<AnkoTilesActivity>(AnkoTilesActivity::class.java)

	private val activity: Activity
		get() = testRule.activity

	private val contentView: View
		get() = activity.findViewById(android.R.id.content)

	private val width: Int
		get() = contentView.width

	private val height: Int
		get() = contentView.height

	private val horizontalPadding: Int
		get() = activity.dip(8)

	private val verticalPadding: Int
		get() = activity.dip(16)

	private val innerHorizontalPadding: Int
		get() = activity.dip(12)

	private val innerVerticalPadding: Int
		get() = activity.dip(24)

	@Test
	@Throws(Exception::class)
	fun layoutShouldBeProperlyInitialized() {
		onView(withClassName(containsString(TilesLayout::class.java.simpleName)))
				.check { view, _ ->
					val layout = view as TilesLayout
					assertThat("layout is a TilesLayout", layout, isA(TilesLayout::class.java))
					assertThat("layout columns", layout.colsCount, equalTo(2))
					assertThat("layout rows", layout.rowsCount, equalTo(3))
					assertThat("layout left padding", layout.paddingLeft, equalTo(horizontalPadding))
					assertThat("layout right padding", layout.paddingRight, equalTo(horizontalPadding))
					assertThat("layout top padding", layout.paddingTop, equalTo(verticalPadding))
					assertThat("layout bottom padding", layout.paddingBottom, equalTo(verticalPadding))
					assertThat("layout inner horizontal padding", layout.innerHorizontalPadding, equalTo(innerHorizontalPadding))
					assertThat("layout inner vertical padding", layout.innerVerticalPadding, equalTo(innerVerticalPadding))
				}
	}

	@Test
	@Throws(Exception::class)
	fun layoutShouldFillActivity() {
		onView(withClassName(containsString(TilesLayout::class.java.simpleName)))
				.check { view, _ ->
					val layout = view as TilesLayout
					assertThat("layout width", layout.width, equalTo(width))
					assertThat("layout height", layout.height, equalTo(height))
				}
	}

	@Test
	@Throws(Exception::class)
	fun firstViewShouldBeTopLeft() {
		onView(withText("Button 1"))
				.check(hasBounds(
						horizontalPadding,
						verticalPadding,
						width / 2 - innerHorizontalPadding / 2,
						(height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 + verticalPadding))
	}

	@Test
	@Throws(Exception::class)
	fun secondViewShouldBeTopRight() {
		onView(withText("Button 2"))
				.check(hasBounds(
						width / 2 + innerHorizontalPadding / 2,
						verticalPadding,
						width - horizontalPadding,
						(height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 + verticalPadding))
	}

	@Test
	@Throws(Exception::class)
	fun thirdViewShouldBeMiddleLeft() {
		onView(withText("Button 3"))
				.check(hasBounds(
						horizontalPadding,
						(height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 + verticalPadding + innerVerticalPadding,
						width / 2 - innerHorizontalPadding / 2,
						height - (height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 - verticalPadding - innerVerticalPadding))
	}

	@Test
	@Throws(Exception::class)
	fun fourthViewShouldBeMiddleRight() {
		onView(withText("Button 4"))
				.check(hasBounds(
						width / 2 + innerHorizontalPadding / 2,
						(height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 + verticalPadding + innerVerticalPadding,
						width - horizontalPadding,
						height - (height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 - verticalPadding - innerVerticalPadding))
	}

	@Test
	@Throws(Exception::class)
	fun fifthViewShouldBeBottomLeft() {
		onView(withText("Button 5"))
				.check(hasBounds(
						horizontalPadding,
						height - (height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 - verticalPadding,
						width / 2 - innerHorizontalPadding / 2,
						height - verticalPadding))
	}

	@Test
	@Throws(Exception::class)
	fun sixthViewShouldBeBottomRight() {
		onView(withText("Button 6"))
				.check(hasBounds(
						width / 2 + innerHorizontalPadding / 2,
						height - (height - 2 * verticalPadding - 2 * innerVerticalPadding) / 3 - verticalPadding,
						width - horizontalPadding,
						height - verticalPadding))
	}
}
